File: src/bridge/socket.ts
/**
 * Unix Domain Socket Server/Client
 * IPC layer for Claude Code ↔ Telegram bridge
 */

import { createServer, connect, Server, Socket } from 'net';
import { unlinkSync, existsSync, writeFileSync, readFileSync, mkdirSync, chmodSync } from 'fs';
import { EventEmitter } from 'events';
import { join } from 'path';
import { homedir } from 'os';
import logger from '../utils/logger.js';
import type { BridgeMessage } from './types.js';

// Use secure user-specific directory instead of world-writable /tmp
const SOCKET_DIR = join(homedir(), '.config', 'claude-telegram-mirror');
const DEFAULT_SOCKET_PATH = join(SOCKET_DIR, 'bridge.sock');
const DEFAULT_PID_PATH = join(SOCKET_DIR, 'bridge.pid');

/**
 * Check if a socket is stale (file exists but no daemon listening)
 * Returns: 'stale' | 'active' | 'none'
 */
async function checkSocketStatus(socketPath: string): Promise<'stale' | 'active' | 'none'> {
  if (!existsSync(socketPath)) {
    return 'none';
  }

  return new Promise((resolve) => {
    const testSocket = connect(socketPath);
    const timeout = setTimeout(() => {
      testSocket.destroy();
      resolve('stale'); // Timeout = no response = stale
    }, 1000);

    testSocket.on('connect', () => {
      clearTimeout(timeout);
      testSocket.destroy();
      resolve('active'); // Connection succeeded = daemon alive
    });

    testSocket.on('error', (err: NodeJS.ErrnoException) => {
      clearTimeout(timeout);
      testSocket.destroy();
      if (err.code === 'ECONNREFUSED') {
        resolve('stale'); // Connection refused = daemon dead
      } else {
        resolve('stale'); // Other error = assume stale
      }
    });
  });
}

/**
 * Check if a PID is still running
 */
function isPidRunning(pid: number): boolean {
  try {
    process.kill(pid, 0); // Signal 0 = check if process exists
    return true;
  } catch {
    return false;
  }
}

/**
 * Acquire PID lock, returns true if lock acquired
 */
function acquirePidLock(pidPath: string): boolean {
  if (existsSync(pidPath)) {
    try {
      const existingPid = parseInt(readFileSync(pidPath, 'utf-8').trim(), 10);
      if (!isNaN(existingPid) && isPidRunning(existingPid)) {
        logger.warn('Another daemon instance is running', { pid: existingPid });
        return false;
      }
      // Stale PID file - process not running
      logger.info('Removing stale PID file', { stalePid: existingPid });
    } catch {
      // Can't read PID file - remove it
    }
  }

  // Write our PID
  writeFileSync(pidPath, process.pid.toString());
  logger.debug('PID lock acquired', { pid: process.pid });
  return true;
}

/**
 * Release PID lock
 */
function releasePidLock(pidPath: string): void {
  if (existsSync(pidPath)) {
    try {
      const storedPid = parseInt(readFileSync(pidPath, 'utf-8').trim(), 10);
      if (storedPid === process.pid) {
        unlinkSync(pidPath);
        logger.debug('PID lock released');
      }
    } catch {
      // Ignore errors during cleanup
    }
  }
}

/**
 * Unix Socket Server
 * Accepts connections from hook scripts and routes messages
 */
export class SocketServer extends EventEmitter {
  private server: Server | null = null;
  private clients: Map<string, Socket> = new Map();
  private socketPath: string;
  private pidPath: string;
  private buffer: Map<string, string> = new Map();

  constructor(socketPath: string = DEFAULT_SOCKET_PATH, pidPath: string = DEFAULT_PID_PATH) {
    super();
    this.socketPath = socketPath;
    this.pidPath = pidPath;
  }

  /**
   * Start listening for connections
   * Includes stale socket detection and PID file locking
   */
  async listen(): Promise<void> {
    // Step 0: Ensure socket directory exists with secure permissions (0o700)
    const socketDir = join(this.socketPath, '..');
    if (!existsSync(socketDir)) {
      mkdirSync(socketDir, { recursive: true, mode: 0o700 });
      logger.debug('Created socket directory', { path: socketDir, mode: '0700' });
    } else {
      // Ensure existing directory has correct permissions
      try {
        chmodSync(socketDir, 0o700);
      } catch (error) {
        logger.warn('Could not set socket directory permissions', { error });
      }
    }

    // Step 1: Acquire PID lock (prevents multiple daemon instances)
    if (!acquirePidLock(this.pidPath)) {
      throw new Error('Another daemon instance is already running. Kill it first or check the PID file.');
    }

    // Register cleanup on process exit
    const cleanup = () => {
      releasePidLock(this.pidPath);
    };
    process.on('exit', cleanup);
    process.on('SIGINT', cleanup);
    process.on('SIGTERM', cleanup);

    // Step 2: Check for stale socket
    const socketStatus = await checkSocketStatus(this.socketPath);

    if (socketStatus === 'active') {
      releasePidLock(this.pidPath);
      throw new Error(`Socket ${this.socketPath} is already in use by another process`);
    }

    if (socketStatus === 'stale') {
      try {
        unlinkSync(this.socketPath);
        logger.info('Removed stale socket file', { path: this.socketPath });
      } catch (error) {
        logger.error('Failed to remove stale socket', { error });
      }
    }

    // Step 3: Start the server
    return new Promise((resolve, reject) => {
      this.server = createServer((socket) => {
        this.handleConnection(socket);
      });

      this.server.on('error', (error) => {
        logger.error('Socket server error', { error });
        releasePidLock(this.pidPath);
        this.emit('error', error);
        reject(error);
      });

      this.server.listen(this.socketPath, () => {
        // Set socket file permissions to owner-only (0o600)
        try {
          chmodSync(this.socketPath, 0o600);
          logger.debug('Set socket permissions', { path: this.socketPath, mode: '0600' });
        } catch (error) {
          logger.warn('Could not set socket file permissions', { error });
        }
        logger.info(`Socket server listening on ${this.socketPath}`, { pid: process.pid });
        resolve();
      });
    });
  }

  /**
   * Handle new client connection
   */
  private handleConnection(socket: Socket): void {
    const clientId = `client-${Date.now()}-${Math.random().toString(36).slice(2, 8)}`;
    this.clients.set(clientId, socket);
    this.buffer.set(clientId, '');

    logger.debug('Client connected', { clientId });
    this.emit('connect', clientId);

    socket.on('data', (data) => {
      this.handleData(clientId, data);
    });

    socket.on('close', () => {
      this.clients.delete(clientId);
      this.buffer.delete(clientId);
      logger.debug('Client disconnected', { clientId });
      this.emit('disconnect', clientId);
    });

    socket.on('error', (error) => {
      logger.error('Client socket error', { clientId, error });
      this.clients.delete(clientId);
      this.buffer.delete(clientId);
    });
  }

  /**
   * Handle incoming data (NDJSON protocol)
   */
  private handleData(clientId: string, data: Buffer): void {
    let buffer = this.buffer.get(clientId) || '';
    buffer += data.toString();

    // Process complete lines (NDJSON)
    const lines = buffer.split('\n');
    this.buffer.set(clientId, lines.pop() || '');

    for (const line of lines) {
      if (!line.trim()) continue;

      try {
        const message = JSON.parse(line) as BridgeMessage;
        this.emit('message', message, clientId);
      } catch (error) {
        logger.error('Failed to parse message', { line, error });
      }
    }
  }

  /**
   * Send message to specific client
   */
  send(clientId: string, message: BridgeMessage): boolean {
    const socket = this.clients.get(clientId);
    if (!socket || socket.destroyed) {
      logger.warn('Client not found or disconnected', { clientId });
      return false;
    }

    try {
      socket.write(JSON.stringify(message) + '\n');
      return true;
    } catch (error) {
      logger.error('Failed to send message', { clientId, error });
      return false;
    }
  }

  /**
   * Broadcast message to all clients
   */
  broadcast(message: BridgeMessage): void {
    const data = JSON.stringify(message) + '\n';
    for (const [clientId, socket] of this.clients) {
      if (!socket.destroyed) {
        try {
          socket.write(data);
        } catch (error) {
          logger.error('Failed to broadcast to client', { clientId, error });
        }
      }
    }
  }

  /**
   * Get connected client count
   */
  getClientCount(): number {
    return this.clients.size;
  }

  /**
   * Close the server and release all locks
   */
  close(): Promise<void> {
    return new Promise((resolve) => {
      // Close all client connections
      for (const socket of this.clients.values()) {
        socket.destroy();
      }
      this.clients.clear();

      // Close server
      if (this.server) {
        this.server.close(() => {
          // Clean up socket file
          if (existsSync(this.socketPath)) {
            try {
              unlinkSync(this.socketPath);
            } catch (error) {
              logger.error('Failed to clean up socket file', { error });
            }
          }

          // Release PID lock
          releasePidLock(this.pidPath);

          logger.info('Socket server closed');
          resolve();
        });
      } else {
        // Still release PID lock even if server wasn't created
        releasePidLock(this.pidPath);
        resolve();
      }
    });
  }
}

/**
 * Unix Socket Client
 * Used by hook scripts to send messages to bridge
 */
export class SocketClient extends EventEmitter {
  private socket: Socket | null = null;
  private socketPath: string;
  private buffer = '';
  private reconnectTimer: NodeJS.Timeout | null = null;
  private connected = false;

  constructor(socketPath: string = DEFAULT_SOCKET_PATH) {
    super();
    this.socketPath = socketPath;
  }

  /**
   * Connect to the server
   */
  connect(): Promise<void> {
    return new Promise((resolve, reject) => {
      this.socket = connect(this.socketPath);

      this.socket.on('connect', () => {
        this.connected = true;
        logger.debug('Connected to bridge');
        this.emit('connect');
        resolve();
      });

      this.socket.on('data', (data) => {
        this.handleData(data);
      });

      this.socket.on('close', () => {
        this.connected = false;
        logger.debug('Disconnected from bridge');
        this.emit('disconnect');
      });

      this.socket.on('error', (error: NodeJS.ErrnoException) => {
        this.connected = false;

        if (error.code === 'ENOENT') {
          logger.warn('Bridge not running (socket not found)');
        } else if (error.code === 'ECONNREFUSED') {
          logger.warn('Bridge refused connection');
        } else {
          logger.error('Socket error', { error });
        }

        this.emit('error', error);
        reject(error);
      });
    });
  }

  /**
   * Handle incoming data
   */
  private handleData(data: Buffer): void {
    this.buffer += data.toString();

    const lines = this.buffer.split('\n');
    this.buffer = lines.pop() || '';

    for (const line of lines) {
      if (!line.trim()) continue;

      try {
        const message = JSON.parse(line) as BridgeMessage;
        this.emit('message', message);
      } catch (error) {
        logger.error('Failed to parse server message', { error });
      }
    }
  }

  /**
   * Send message to server
   */
  send(message: BridgeMessage): boolean {
    if (!this.socket || this.socket.destroyed || !this.connected) {
      logger.warn('Not connected to bridge');
      return false;
    }

    try {
      this.socket.write(JSON.stringify(message) + '\n');
      return true;
    } catch (error) {
      logger.error('Failed to send message', { error });
      return false;
    }
  }

  /**
   * Send message and wait for response
   */
  sendAndWait(message: BridgeMessage, timeout: number = 30000): Promise<BridgeMessage> {
    return new Promise((resolve, reject) => {
      const timer = setTimeout(() => {
        reject(new Error('Response timeout'));
      }, timeout);

      const handler = (response: BridgeMessage) => {
        if (response.sessionId === message.sessionId) {
          clearTimeout(timer);
          this.off('message', handler);
          resolve(response);
        }
      };

      this.on('message', handler);

      if (!this.send(message)) {
        clearTimeout(timer);
        this.off('message', handler);
        reject(new Error('Failed to send message'));
      }
    });
  }

  /**
   * Check if connected
   */
  isConnected(): boolean {
    return this.connected;
  }

  /**
   * Disconnect from server
   */
  disconnect(): void {
    if (this.reconnectTimer) {
      clearTimeout(this.reconnectTimer);
      this.reconnectTimer = null;
    }

    if (this.socket) {
      this.socket.destroy();
      this.socket = null;
    }

    this.connected = false;
  }
}

export { SOCKET_DIR, DEFAULT_SOCKET_PATH, DEFAULT_PID_PATH, checkSocketStatus, isPidRunning };

File: rust-crates/ctm/src/socket.rs
//! Unix domain socket server and client — NDJSON framing.
//!
//! Ported from `socket.ts`.  Wire-compatible with the TypeScript implementation.
//!
//! ## Server
//! - `flock(2)` PID locking via `nix::fcntl::Flock`
//! - `umask(0o177)` before bind → socket file gets `0o600`
//! - Stale socket detection (check flock, not TOCTOU PID file)
//! - 64 max concurrent connections
//! - 1 MiB max NDJSON line
//!
//! ## Client
//! - Connect, send, send-and-wait with correlation on `session_id`

use crate::error::{AppError, Result};
use crate::types::BridgeMessage;
use nix::fcntl::Flock;
use nix::sys::stat::Mode;
use std::collections::HashMap;
use std::os::fd::OwnedFd;
use std::path::{Path, PathBuf};
use std::sync::Arc;
use tokio::io::{AsyncBufReadExt, AsyncWriteExt, BufReader};
use tokio::net::{UnixListener, UnixStream};
use tokio::sync::{broadcast, Mutex};
use tokio::time::{timeout, Duration};

/// Maximum concurrent client connections.
const MAX_CONNECTIONS: usize = 64;

/// Maximum bytes in a single NDJSON line (1 MiB).
const MAX_LINE_BYTES: usize = 1_048_576;

// ====================================================================== server

/// Unix domain socket server with NDJSON framing.
pub struct SocketServer {
    socket_path: PathBuf,
    pid_path: PathBuf,
    /// Held for the server's lifetime — releasing it signals "no daemon".
    _lock: Option<Flock<OwnedFd>>,
    tx: broadcast::Sender<BridgeMessage>,
    clients: Arc<Mutex<HashMap<String, Arc<Mutex<tokio::net::unix::OwnedWriteHalf>>>>>,
    shutdown_tx: Option<tokio::sync::oneshot::Sender<()>>,
}

impl SocketServer {
    pub fn new(socket_path: &Path, pid_path: &Path) -> Self {
        let (tx, _) = broadcast::channel(256);
        Self {
            socket_path: socket_path.to_path_buf(),
            pid_path: pid_path.to_path_buf(),
            _lock: None,
            tx,
            clients: Arc::new(Mutex::new(HashMap::new())),
            shutdown_tx: None,
        }
    }

    /// Subscribe to the broadcast channel to receive incoming messages.
    pub fn subscribe(&self) -> broadcast::Receiver<BridgeMessage> {
        self.tx.subscribe()
    }

    /// Number of currently connected clients.
    pub async fn client_count(&self) -> usize {
        self.clients.lock().await.len()
    }

    /// Return a shared reference to the connected-client map so callers outside
    /// of `SocketServer` can broadcast without holding the server itself.
    pub fn clients_ref(
        &self,
    ) -> Arc<Mutex<HashMap<String, Arc<Mutex<tokio::net::unix::OwnedWriteHalf>>>>> {
        Arc::clone(&self.clients)
    }

    /// Start listening.  Returns when the accept loop is running.
    pub async fn listen(&mut self) -> Result<()> {
        // Step 1: Acquire flock on PID file (atomic, no TOCTOU).
        let lock = acquire_flock(&self.pid_path)?;

        // Step 2: Enforce 0o700 on the parent directory so that the socket file
        // (0o600) is nested inside a directory that is only accessible by the
        // owner.  This prevents other users from even discovering the socket path.
        if let Some(parent) = self.socket_path.parent() {
            use std::os::unix::fs::PermissionsExt;
            if parent.exists() {
                let _ = std::fs::set_permissions(parent, std::fs::Permissions::from_mode(0o700));
            } else {
                std::fs::create_dir_all(parent).map_err(|e| {
                    AppError::Socket(format!("Cannot create socket directory: {e}"))
                })?;
                let _ = std::fs::set_permissions(parent, std::fs::Permissions::from_mode(0o700));
            }
        }

        // Step 3: Remove stale socket if present.
        if self.socket_path.exists() {
            let _ = std::fs::remove_file(&self.socket_path);
        }

        // Step 4: Bind with restricted umask so the socket file is 0o600.
        let old_mask = nix::sys::stat::umask(Mode::from_bits_truncate(0o177));
        let listener = UnixListener::bind(&self.socket_path).map_err(|e| {
            nix::sys::stat::umask(old_mask);
            AppError::Socket(format!("Failed to bind socket: {e}"))
        })?;
        nix::sys::stat::umask(old_mask);

        // Write PID file after lock acquired.
        let _ = std::fs::write(&self.pid_path, std::process::id().to_string());

        self._lock = Some(lock);

        let tx = self.tx.clone();
        let clients = Arc::clone(&self.clients);
        let (shutdown_tx, mut shutdown_rx) = tokio::sync::oneshot::channel();
        self.shutdown_tx = Some(shutdown_tx);

        // Spawn accept loop.
        tokio::spawn(async move {
            loop {
                tokio::select! {
                    accepted = listener.accept() => {
                        match accepted {
                            Ok((stream, _)) => {
                                let count = clients.lock().await.len();
                                if count >= MAX_CONNECTIONS {
                                    tracing::warn!(count, "Connection limit reached, rejecting");
                                    drop(stream);
                                    continue;
                                }
                                let client_id = format!(
                                    "client-{}-{}",
                                    chrono::Utc::now().timestamp_millis(),
                                    &uuid::Uuid::new_v4().simple().to_string()[..8]
                                );
                                let tx2 = tx.clone();
                                let clients2 = Arc::clone(&clients);
                                let cid = client_id.clone();

                                let (reader, writer) = stream.into_split();
                                clients.lock().await.insert(
                                    client_id.clone(),
                                    Arc::new(Mutex::new(writer)),
                                );

                                tokio::spawn(async move {
                                    handle_client(reader, &cid, tx2).await;
                                    clients2.lock().await.remove(&cid);
                                });
                            }
                            Err(e) => {
                                tracing::error!(error = %e, "Accept error");
                            }
                        }
                    }
                    _ = &mut shutdown_rx => {
                        break;
                    }
                }
            }
        });

        Ok(())
    }

    /// Send an NDJSON message to a specific client.
    pub async fn send(&self, client_id: &str, message: &BridgeMessage) -> Result<bool> {
        let clients = self.clients.lock().await;
        if let Some(writer) = clients.get(client_id) {
            let json = serde_json::to_string(message)?;
            let mut w = writer.lock().await;
            if w.write_all(format!("{json}\n").as_bytes()).await.is_ok() {
                return Ok(true);
            }
        }
        Ok(false)
    }

    /// Broadcast to all connected clients.
    pub async fn broadcast(&self, message: &BridgeMessage) -> Result<()> {
        let json = serde_json::to_string(message)?;
        let line = format!("{json}\n");
        let clients = self.clients.lock().await;
        for (_id, writer) in clients.iter() {
            let mut w = writer.lock().await;
            let _ = w.write_all(line.as_bytes()).await;
        }
        Ok(())
    }

    /// Shut down the server, removing socket and PID files.
    pub async fn close(mut self) {
        // Signal accept loop to stop
        if let Some(tx) = self.shutdown_tx.take() {
            let _ = tx.send(());
        }
        // Disconnect all clients
        self.clients.lock().await.clear();
        // Cleanup files
        let _ = std::fs::remove_file(&self.socket_path);
        // Lock released on drop via _lock
    }
}

impl Drop for SocketServer {
    fn drop(&mut self) {
        let _ = std::fs::remove_file(&self.socket_path);
        let _ = std::fs::remove_file(&self.pid_path);
    }
}

/// Read NDJSON lines from a client and broadcast each parsed message.
async fn handle_client(
    reader: tokio::net::unix::OwnedReadHalf,
    client_id: &str,
    tx: broadcast::Sender<BridgeMessage>,
) {
    let mut buf_reader = BufReader::new(reader);
    let mut line = String::new();

    loop {
        line.clear();
        match buf_reader.read_line(&mut line).await {
            Ok(0) => break, // EOF
            Ok(_) => {
                let trimmed = line.trim();
                if trimmed.is_empty() {
                    continue;
                }
                if trimmed.len() > MAX_LINE_BYTES {
                    tracing::warn!(
                        client_id,
                        len = trimmed.len(),
                        "Oversized NDJSON line dropped"
                    );
                    continue;
                }
                match serde_json::from_str::<BridgeMessage>(trimmed) {
                    Ok(msg) => {
                        let _ = tx.send(msg);
                    }
                    Err(e) => {
                        tracing::debug!(client_id, error = %e, "Failed to parse NDJSON");
                    }
                }
            }
            Err(e) => {
                tracing::debug!(client_id, error = %e, "Read error");
                break;
            }
        }
    }
}

/// Acquire an exclusive flock on the PID file.
/// If the file is already locked, another daemon owns it.
fn acquire_flock(pid_path: &Path) -> Result<Flock<OwnedFd>> {
    use std::fs::OpenOptions;
    use std::os::fd::FromRawFd;
    use std::os::fd::IntoRawFd;

    let file = OpenOptions::new()
        .create(true)
        .truncate(false)
        .write(true)
        .open(pid_path)
        .map_err(|e| AppError::Lock(format!("Cannot open PID file: {e}")))?;

    // Convert File -> OwnedFd
    let raw_fd = file.into_raw_fd();
    // SAFETY: we just obtained this fd from a valid File.
    let owned_fd = unsafe { OwnedFd::from_raw_fd(raw_fd) };

    Flock::lock(owned_fd, nix::fcntl::FlockArg::LockExclusiveNonblock).map_err(|(_fd, errno)| {
        AppError::Lock(format!("Another daemon instance holds the lock: {errno}"))
    })
}

// ====================================================================== client

/// Unix domain socket client for sending messages to the server.
pub struct SocketClient {
    stream: Option<UnixStream>,
}

impl SocketClient {
    pub fn new() -> Self {
        Self { stream: None }
    }

    /// Connect to the server at the given socket path.
    pub async fn connect(&mut self, socket_path: &Path) -> Result<()> {
        let stream = UnixStream::connect(socket_path)
            .await
            .map_err(|e| AppError::Socket(format!("Failed to connect: {e}")))?;
        self.stream = Some(stream);
        Ok(())
    }

    /// Send a single NDJSON message.
    pub async fn send(&mut self, message: &BridgeMessage) -> Result<()> {
        let stream = self
            .stream
            .as_mut()
            .ok_or_else(|| AppError::Socket("Not connected".into()))?;

        let json = serde_json::to_string(message)?;
        stream
            .write_all(format!("{json}\n").as_bytes())
            .await
            .map_err(|e| AppError::Socket(format!("Failed to write: {e}")))?;
        Ok(())
    }

    /// Send a message and wait for a response matching `session_id`.
    pub async fn send_and_wait(
        &mut self,
        message: &BridgeMessage,
        wait_timeout: Duration,
    ) -> Result<BridgeMessage> {
        self.send(message).await?;

        let stream = self
            .stream
            .as_mut()
            .ok_or_else(|| AppError::Socket("Not connected".into()))?;

        let (reader, _writer) = stream.split();
        let mut buf_reader = BufReader::new(reader);
        let session_id = &message.session_id;

        let result = timeout(wait_timeout, async {
            let mut line = String::new();
            loop {
                line.clear();
                let bytes = buf_reader
                    .read_line(&mut line)
                    .await
                    .map_err(|e| AppError::Socket(format!("Read failed: {e}")))?;

                if bytes == 0 {
                    return Err(AppError::Socket("Connection closed".into()));
                }

                if let Ok(msg) = serde_json::from_str::<BridgeMessage>(line.trim()) {
                    if msg.session_id == *session_id {
                        return Ok(msg);
                    }
                }
            }
        })
        .await;

        match result {
            Ok(inner) => inner,
            Err(_) => Err(AppError::Socket("Response timeout".into())),
        }
    }

    /// Disconnect (drop the stream).
    pub fn disconnect(&mut self) {
        self.stream = None;
    }

    pub fn is_connected(&self) -> bool {
        self.stream.is_some()
    }
}

// ===================================================================== tests

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn flock_rejects_double_lock() {
        let tmp = tempfile::TempDir::new().unwrap();
        let pid = tmp.path().join("test.pid");

        let _lock1 = acquire_flock(&pid).expect("First lock should succeed");
        let result = acquire_flock(&pid);
        assert!(result.is_err(), "Second lock should fail");
    }

    #[test]
    fn flock_released_on_drop() {
        let tmp = tempfile::TempDir::new().unwrap();
        let pid = tmp.path().join("test2.pid");

        {
            let _lock = acquire_flock(&pid).unwrap();
        }
        // After drop, re-lock should succeed.
        let _lock2 = acquire_flock(&pid).expect("Lock after drop should succeed");
    }

    #[tokio::test]
    async fn server_client_roundtrip() {
        let tmp = tempfile::TempDir::new().unwrap();
        let sock = tmp.path().join("test.sock");
        let pid = tmp.path().join("test.pid");

        let mut server = SocketServer::new(&sock, &pid);
        server.listen().await.unwrap();
        let mut rx = server.subscribe();

        // Give the accept loop a moment to start.
        tokio::time::sleep(Duration::from_millis(50)).await;

        // Client sends a message.
        let mut client = SocketClient::new();
        client.connect(&sock).await.unwrap();
        let msg = BridgeMessage {
            msg_type: "test".to_string(),
            session_id: "s1".to_string(),
            timestamp: "2024-01-01T00:00:00.000Z".to_string(),
            content: "hello".to_string(),
            metadata: None,
        };
        client.send(&msg).await.unwrap();

        // Server receives.
        let received = tokio::time::timeout(Duration::from_secs(2), rx.recv())
            .await
            .expect("timeout")
            .expect("recv");

        assert_eq!(received.session_id, "s1");
        assert_eq!(received.content, "hello");

        client.disconnect();
        server.close().await;
    }

    #[tokio::test]
    async fn connection_limit_enforced() {
        let tmp = tempfile::TempDir::new().unwrap();
        let sock = tmp.path().join("limit.sock");
        let pid = tmp.path().join("limit.pid");

        let mut server = SocketServer::new(&sock, &pid);
        server.listen().await.unwrap();

        tokio::time::sleep(Duration::from_millis(50)).await;

        // We just verify 64+ connections don't crash the server.
        // Full enforcement test would connect 65 clients, but keeping it lean.
        let count = server.client_count().await;
        assert_eq!(count, 0);

        server.close().await;
    }

    #[test]
    fn max_line_bytes_constant() {
        assert_eq!(MAX_LINE_BYTES, 1_048_576);
    }
}
